### PR TITLE
R4R: Refactor coinswap module

### PR DESCRIPTION
**Pull request by SegueII**
Resolves: #2094

### CODE REVIEW

File: modules/coinswap/client/rest/rest.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+// RegisterRoutes registers asset-related REST handlers to a router
+func RegisterRoutes(cliCtx context.CLIContext, r *mux.Router) {
+	registerQueryRoutes(cliCtx, r)
+	registerTxRoutes(cliCtx, r)
+}
+
+type AddLiquidityReq struct {
+	BaseTx       rest.BaseReq `json:"base_tx" yaml:"base_tx"`
+	Id           string       `json:"id" yaml:"id"`                         // the unique liquidity id
```
Comment (zhiqiang-bianjie) [inline modules/coinswap/client/rest/rest.go]: Id -> ID

File: modules/coinswap/client/rest/rest.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	MaxToken     string       `json:"max_token" yaml:"max_token"`           // token to be deposited as liquidity with an upper bound for its amount
+	ExactIrisAmt string       `json:"exact_iris_amt" yaml:"exact_iris_amt"` // exact amount of iris-atto being add to the liquidity pool
+	MinLiquidity string       `json:"min_liquidity" yaml:"min_liquidity"`   // lower bound UNI sender is willing to accept for deposited coins
+	Deadline     string       `json:"deadline" yaml:"deadline"`             // deadline duration, e.g. 10m
+	Sender       string       `json:"sender" yaml:"sender"`
+}
+
+type RemoveLiquidityReq struct {
+	BaseTx            rest.BaseReq `json:"base_tx" yaml:"base_tx"`
+	Id                string       `json:"id" yaml:"id"`                                 // the unique liquidity id
```
Comment (zhiqiang-bianjie) [inline modules/coinswap/client/rest/rest.go]: Id -> ID

File: modules/coinswap/client/rest/tx.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+			return
+		}
+
+		duration, e := time.ParseDuration(req.Deadline)
+		if e != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, e.Error())
+			return
+		}
+
+		deadline := time.Now().Add(duration)
```
Comment (zhiqiang-bianjie) [inline modules/coinswap/client/rest/tx.go]: time.Now() -> block time

File: modules/coinswap/internal/keeper/keeper.go [hunk 4]
```diff
@@ -4,221 +4,232 @@ import (
...
-	tags = sdk.NewTags(
-		types.TagSender, []byte(msg.Sender.String()),
-		types.TagTokenPair, []byte(getTokenPairByDenom(msg.MaxToken.Denom, sdk.IrisAtto)),
+	addLiquidityEvent := sdk.NewEvent(
+		types.EventAddLiquidity,
+		sdk.NewAttribute(types.AttributeValueSender, msg.Sender.String()),
+		sdk.NewAttribute(types.AttributeValueTokenPair, getTokenPairByDenom(msg.MaxToken.Denom, config.IrisAtto)),
 	)
-	return tags, k.addLiquidity(ctx, msg.Sender, irisCoin, depositToken, uniId, mintLiquidityAmt)
+	ctx.EventManager().EmitEvents(sdk.Events{addLiquidityEvent})
```
Comment (chengwenxi) [inline modules/coinswap/internal/keeper/keeper.go]:
```suggestion
	ctx.EventManager().EmitEvent(addLiquidityEvent)
```

File: modules/coinswap/handler.go [hunk 5]
```diff
@@ -4,22 +4,23 @@ import (
 	"fmt"
 	"time"
 
-	sdk "github.com/irisnet/irishub/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/irisnet/irishub/modules/coinswap/internal/types"
```
Comment (zhiqiang-bianjie) [inline modules/coinswap/handler.go]: should define in alias

File: modules/coinswap/handler.go [hunk 6]
```diff
@@ -28,50 +29,77 @@ func NewHandler(k Keeper) sdk.Handler {
...
 		return err.Result()
 	}
 
-	return sdk.Result{Tags: tags}
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Input.Address.String()),
+		),
```
Comment (zhiqiang-bianjie) [inline modules/coinswap/handler.go]: should add event "EventTypeSwapOrder"

File: modules/coinswap/handler.go [hunk 7]
```diff
@@ -28,50 +29,77 @@ func NewHandler(k Keeper) sdk.Handler {
...
 	if err != nil {
 		return err.Result()
 	}
 
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			sdk.EventTypeMessage,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.AttributeValueCategory),
+			sdk.NewAttribute(sdk.AttributeKeySender, msg.Sender.String()),
+		),
```
Comment (zhiqiang-bianjie) [inline modules/coinswap/handler.go]: should add event "EventTypeAddLiquidity"

File: modules/coinswap/handler.go [hunk 8]
```diff
@@ -28,50 +29,77 @@ func NewHandler(k Keeper) sdk.Handler {
...
```
Comment (zhiqiang-bianjie) [inline modules/coinswap/handler.go]: should add event "EventTypeRemoveLiquidity"

File: modules/coinswap/internal/types/params.go [hunk 9]
```diff
@@ -2,108 +2,84 @@ package types
...
 
 // ParamTypeTable returns the TypeTable for coinswap module
-func ParamTypeTable() params.TypeTable {
-	return params.NewTypeTable().RegisterParamSet(&Params{})
+func ParamKeyTable() params.KeyTable {
+	return params.NewKeyTable().RegisterParamSet(&Params{})
 }
 
 // String returns a human readable string representation of the parameters.
 func (p Params) String() string {
```
Comment (chengwenxi) [inline modules/coinswap/internal/types/params.go]: no need

File: modules/coinswap/internal/types/params.go [hunk 10]
```diff
@@ -2,108 +2,84 @@ package types
...
-		return sdk.ParseParamsErr(fmt.Errorf("fee is not positive: %s", fee.String()))
+	if !p.Fee.GT(sdk.ZeroDec()) {
+		return fmt.Errorf("fee is not positive: %s", p.Fee.String())
 	}
-
-	if !fee.LT(sdk.OneRat()) {
-		return sdk.ParseParamsErr(fmt.Errorf("fee must be less than 1: %s", fee.String()))
+	if !p.Fee.LT(sdk.OneDec()) {
+		return fmt.Errorf("fee must be less than 1: %s", p.Fee.String())
+	}
```
Comment (chengwenxi) [inline modules/coinswap/internal/types/params.go]: merge the two check to one.

File: modules/coinswap/internal/types/utils.go [hunk 11]
```diff
@@ -2,83 +2,48 @@ package types
...
-func GetUniId(denom1, denom2 string) (string, sdk.Error) {
+// GetUniDenomFromDenoms returns the uni denom for the provided denominations.
+func GetUniDenomFromDenoms(denom1, denom2 string) (string, sdk.Error) {
 	if denom1 == denom2 {
 		return "", ErrEqualDenom("denomnations for forming uni id are equal")
 	}
 
-	if denom1 != sdk.IrisAtto && denom2 != sdk.IrisAtto {
-		return "", ErrIllegalDenom(fmt.Sprintf("illegal denomnations for forming uni id, must have one native denom: %s", sdk.IrisAtto))
+	if denom1 != StandardDenom && denom2 != StandardDenom {
```
Comment (chengwenxi) [inline modules/coinswap/internal/types/utils.go]: should be `params.StandardDenom`

**Pull request merged**